packages/core/src/sync-realtime/service.ts

```diff
@@ -8,878 +8,633 @@ import {
-      );
-      this.logs = this.logs.filter(
-        (log) => log.blockNumber > newFinalizedBlock.number,
-      );
-
-      // TODO: Update this to insert:
-      // 1) Log filter intervals
-      // 2) Factory contract intervals
-      // 3) Child filter intervals
-      await this.syncStore.insertRealtimeInterval({
-        chainId: this.network.chainId,
-        logFilters: this.logFilterSources.map((l) => l.criteria),
-        factories: this.factorySources.map((f) => f.criteria),
-        interval: {
-          startBlock: BigInt(this.finalizedBlock.number + 1),
-          endBlock: BigInt(newFinalizedBlock.number),
-        },
-      });
+          return;
+        } catch (_error) {
+          if (realtimeSyncService.isKilled) return;
 
-      this.finalizedBlock = newFinalizedBlock;
+          const error = _error as Error;
 
-      this.emit("finalityCheckpoint", {
-        ...maxCheckpoint,
-        blockTimestamp: newFinalizedBlock.timestamp,
-        chainId: this.network.chainId,
-        blockNumber: newFinalizedBlock.number,
-      });
+          realtimeSyncService.common.logger.warn({
+            service: "realtime",
+            msg: `Realtime sync task failed (network=${
+              realtimeSyncService.network.name
+            }, error=${`${error.name}: ${error.message}`})`,
+          });
 
-      this.common.logger.debug({
-        service: "realtime",
-        msg: `Updated finality checkpoint to ${newFinalizedBlock.number} (network=${this.network.name})`,
-      });
-    }
+          if (i === 5) realtimeSyncService.onFatalError(error);
+          else await wait(250 * 2 ** i);
+        }
+      }
+    },
+  });
+
+  const enqueue = () =>
+    _eth_getBlockByNumber(realtimeSyncService, {
+      blockTag: "latest",
+    }).then(queue.add);
+
+  const interval = setInterval(
+    enqueue,
+    realtimeSyncService.network.pollingInterval,
+  );
+
+  realtimeSyncService.kill = async () => {
+    realtimeSyncService.isKilled = true;
+    clearInterval(interval);
+    queue.pause();
+    queue.clear();
+
+    await queue.onIdle();
+  };
 
-    const newBlockNumber = hexToNumber(newBlock.number);
-    const newBlockTimestamp = hexToNumber(newBlock.timestamp);
+  // Note: this is done just for testing.
+  return enqueue().then(() => queue);
+};
 
-    this.emit("realtimeCheckpoint", {
-      ...maxCheckpoint,
-      blockTimestamp: newBlockTimestamp,
-      chainId: this.network.chainId,
-      blockNumber: newBlockNumber,
-    });
+/**
+ * 1) Determine if a reorg occurred.
+ * 2) Insert new event data into the store.
+ * 3) Determine if a new range of events has become finalized,
+ *    if so validate the local chain before removing the
+ *    finalized data.
+ *
+ * @param pendingLatestBlock Block to be injested. Must be exactly
+ * 1 block ahead of the local chain.
+ * @returns true if a reorg occurred
+ */
+export const handleBlock = async (
+  realtimeSyncService: Service,
```

Comment:
You could probably get away with naming this `service` to make this function less verbose.